### ISSUE TITLE
gopackagesdriver: incorrect assumptions around canonical labels

### ISSUE THREAD

**Issue by mikedanese** [bazelbuild/rules_go#3865]
### What version of rules_go are you using?

```
http_archive(
    name = "io_bazel_rules_go",
    sha256 = "6734a719993b1ba4ebe9806e853864395a8d3968ad27f9dd759c196b3eb3abe8",
    urls = [
        "https://mirror.bazel.build/github.com/bazelbuild/rules_go/releases/download/v0.45.1/rules_go-v0.45.1.zip",
        "https://github.com/bazelbuild/rules_go/releases/download/v0.45.1/rules_go-v0.45.1.zip",
    ],
)
```

### What version of Bazel are you using?

```
Build label: 6.4.0
Build target: bazel-out/darwin_arm64-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Thu Oct 19 17:08:20 2023 (1697735300)
Build timestamp: 1697735300
Build timestamp as int: 1697735300
```

### Does this issue reproduce with the latest releases of all the above?

Yes.

### What operating system and processor architecture are you using?

MacCOS Arm64.

### What did you do?

gopls does not work following the editor instructions here: https://github.com/bazelbuild/rules_go/wiki/Editor-and-tool-integration

It does work, however, if I comment out these lines in gopackagesdriver:

https://github.com/bazelbuild/rules_go/blob/f0e46ec3e1edf929ded5fa8fbf9e016fd08e57c8/go/tools/gopackagesdriver/packageregistry.go#L105-L109

Even with bazel 6.4.0, the aspect is still generating package IDs that are not in the new canonical form, e.g.:

```
$ cat bazel-bin/crypto/aesxcbc/aesxcbc.pkg.json  | jq '.'
{
  "ID": "//crypto/aesxcbc:aesxcbc",
  ...
}
```

Only once I switch the repo to using bzlmod do I see IDs generated in canonical form:

```
$ cat bazel-bin/crypto/aesxcbc/aesxcbc.pkg.json | jq '.'
{
  "ID": "@@//crypto/aesxcbc:aesxcbc",
  ...
}
```

### TL;DR

I believe the assumption on when to expect canonical paths is wrong. IIUC, we currently expect them after bazel 6.0.0 however they do not appear to be used unless the repo is also migrated to bzlmod.

**Comment by annapst** [comment]
I'm running into the same issue, I'm also commenting out now! Using bazel 7.1.1.
